pkgs/misc/tmux-plugins/default.nix

```diff
@@ -377,6 +377,17 @@ in rec {
     };
   };
 
+  tmux-fzf = mkDerivation {
+    pluginName = "tmux-fzf";
```

Comment:
:raising_hand:  Since `tmux-fzf` expects `fzf` to be available at runtime, we need to patch *[scripts/.fzf-tmux](https://github.com/sainnhe/tmux-fzf/blob/570f8baff3cb176d7339eb42cddc98e5695458f7/scripts/.fzf-tmux#L20)* to look for the executable in the nix store. You can find examples of this in `postInstall` hooks inside this file.

Reply:
Should I also patch [`sed`](https://github.com/sainnhe/tmux-fzf/search?q=sed)?

Reply:
@KyleOndy excellent suggestion, we should indeed :smiley:

Reply:
I'm happy to do that work; I am just curious how complete this should be? Feel free to reroute me to a more general forum for this if appropriate.

Should I patch all instances on binary calls to use the executable in the nix store? `grep`? `xargs`? The calls to `tmux` itself?

This is where my lack of deeper knowledge of nix come in, and I admit I have not finished RTFM, so if that is the right move, let me know.

If the system is going to resolve the binary to the nix store eventually, what is the reason or patching the script directly? If the purpose is to ensure the binary is available, shouldn't that be resolved with a [dependeny](https://github.com/NixOS/nixpkgs/blob/440c5808a18980909ca2eb349d4819d266967e49/pkgs/misc/tmux-plugins/default.nix#L76)?

----

```
> which fzf
/home/kyle/.nix-profile/bin/fzf

> ls -l /home/kyle/.nix-profile/bin/fzf
lrwxrwxrwx 1 root root 69 Dec 31  1969 /home/kyle/.nix-profile/bin/fzf -> /nix/store/8ppzzf9lnnpdccmm9bdk8miadpvxfwi3-home-manager-path/bin/fzf

> ls -l /nix/store/8ppzzf9lnnpdccmm9bdk8miadpvxfwi3-home-manager-path/bin/fzf
lrwxrwxrwx 1 root root 62 Dec 31  1969 /nix/store/8ppzzf9lnnpdccmm9bdk8miadpvxfwi3-home-manager-path/bin/fzf -> /nix/store/gklqlnxfhx5rrphdx1mjvxmmi597khc8-fzf-0.24.3/bin/fzf
```

Reply:
The aim to is create a derivation that captures all dependencies (at the exception of `tmux` perhaps) to use the package. This can be verified in a hermetic environment provided by `nix-shell --pure`.

I recommend to first checkout the PR with [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review). This will checkout the PR in a temporary directory, build the derivation and spawn an interactive shell.
```sh
# run the command at the root of `nixpkgs` repo
nixpkgs-review pr 95275
```

Inside the temporary directory find `shell.nix` and replace its content with the snippet below. This will provide a shell with tmux, the plugin and a tmux config file.
```nix
{ pkgs ? import ./nixpkgs {} }:
with pkgs;
let
  tmuxConfig = ''
    set -g @plugin 'sainnhe/tmux-fzf'

    run '${tmuxPlugins.tmux-fzf}/share/tmux-plugins/tmux-fzf/main.tmux'
  '';
in stdenv.mkDerivation {
  name = "env";
  buildInputs = [
    tmuxPlugins."tmux-fzf"
    tmux
  ];
  unpackPhase = ":";
  installPhase = "touch $out";
  shellHook = ''
    echo "${tmuxConfig}" > $PWD/tmux.config
  '';
}
```

Then run `nix-shell --pure` (by default will use `shell.nix` in the current directory). Once the shell started, run `tmux -f tmux.config` to test the plugin. There we can observe the plugin is not yet functional.

I know of two options to specify dependencies, 1) patch the source files to reference nix store location 2) prefix the entry point's `PATH` with `wrapProgram` ([example](https://github.com/NixOS/nixpkgs/blob/bc714f86a515a558c2f595e1a090e882d08bd7ca/pkgs/os-specific/linux/alsa-utils/default.nix#L19-L21)).
Based on how the plugin is built, `wrapProgram` could be a neat solution.

Now, onto the existing `dependencies` attribute, it appears to do nothing. I would be looking at deprecating its use.

On this output. It shows `fzf` is installed in your profile, but the plugin should work irrespective of whether it is in your profile.
```
> which fzf
/home/kyle/.nix-profile/bin/fzf
```

Let me know if it does not make sense or you have further questions. I am learning too :nerd_face:

Reply:
Thank you so much for this write up. I appreciate the time it took. This connected a lot of dots for me, and made things much clearer.

> On this output. It shows fzf is installed in your profile, but the plugin should work irrespective of whether it is in your profile.

I don't know why it wasn't obvious to me, but this comment cleared it all up.

----

I'll give this some time tonight and hope to have it all working.

Reply:
I was able to get the plugin working using the following as my `shell.nix`. Notably I dropped the ` set -g @plugin 'sainnhe/tmux-fzf'` line as the `run` command takes care of loading the plugin. The plugin also requires `ncurses`.

```
I was able to get the plugin working using the following as my

```{ pkgs ? import ./nixpkgs {} }:
with pkgs;
let
  tmuxConfig = ''
    run '${tmuxPlugins.tmux-fzf}/share/tmux-plugins/tmux-fzf/main.tmux'
  '';
in stdenv.mkDerivation {
  name = "env";
  buildInputs = [
    tmuxPlugins."tmux-fzf"
    tmux
    ncurses
  ];
  unpackPhase = ":";
  installPhase = "touch $out";
  shellHook = ''
    echo "${tmuxConfig}" > $PWD/tmux.config
  '';
}
```

Reply:
This is looking good @KyleOndy, good spot on the `ncurses` dependency.

When it comes to declaring dependencies, you have effectively done so when referring to `pkgs.<package>` in the substitution steps. This can be observed by looking at derivation's inputs.
```sh
# compute derivation path needed for nix show-derivation
nix-instantiate '<nixpkgs>' -A tmuxPlugins.tmux-fzf
outputs <path to drv>

nix show-derivation <path to drv>
```

In the output find  `inputDrvs` (the derivation's direct dependencies)
```
"inputDrvs": {                                             
  "/nix/store/10xi5iv5svzxfdxdjzbq0hyxwkyxg8dr-bash-4.4-p23.drv": [          
    "out"                                                  
  ],                                                       
  "/nix/store/7bk8andhhffxcb34nhgrgghillmil8xr-stdenv-linux.drv": [
    "out"                                                  
  ],                 
  "/nix/store/jhnqycphsrkcllwwvb31mgbgdw2br3ij-source.drv": [                                                                     "out"                 
  ],
  "/nix/store/sx2ns26na7yxmm3lwnm6lrnin6b9axi6-ncurses-6.2.drv": [
    "out"
  ],
  "/nix/store/ww9by1sif2l0c1608ry4n86scjrgy6vy-fzf-0.24.3.drv": [
    "out"
  ],
  "/nix/store/yq5r43jzh1nl74xs4x9rcxgvd3xqhnck-gnused-4.8.drv": [
    "out"
  ]
},
```